scripts/consensus_merge.py

```diff
@@ -0,0 +1,660 @@
+        output_record.samples[index]['AD'] = get_ad_consensus(AD_list)
+        output_record.samples[index]['AF'] = get_af_consensus(AF_list)
+        output_record.samples[index]['DP'] = get_dp_consensus(DP_list)
+        output_record.samples[index]['ADR'] = (ad_range_1, ad_range_1)
+        output_record.samples[index]['AFR'] = af_range
+        output_record.samples[index]['DPR'] = dp_range
+
+    chrom = single_caller_variants[0].record.chrom
+    pos = single_caller_variants[0].record.pos
+    mapq, mq0 = get_mapq(normal_cram, chrom, pos)
+
+    output_record.info['MQ'] = mapq
+    output_record.info['MQ0'] = mq0
+    output_record.info['CAL'] = ','.join([c for c in CALLER_NAMES if c.lower() in variant_lookup])
+
+    if hotspot:
+        output_record.info['Hotspot'] = True
+
+    output_record.filter.add('PASS')
+
+    output_vcf.write(output_record)
+
+def build_output_name(inpath, outbase):
+    """Builds an VCF.GZ output filename based on the input (VCF/VCF.GZ) name and given output basename
+       Args:
+           inpath (str): Path to the input VCF(.GZ) file
+           outbase (str): Given output basename string
+       Return:
+           str: Filename in the format <output_basename>.<input_nameroot>.consensus.vcf.gz
+    """
+    basename_split = os.path.split(inpath)[-1].split('.')
+    return '.'.join([outbase, basename_split[0] + '.consensus.vcf.gz'])
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--strelka2_vcf', 
+                        help='Strelka2 VCF with MNPs fixed')
+    parser.add_argument('--mutect2_vcf',
+                        help='Mutect2 VCF')
+    parser.add_argument('--lancet_vcf', 
+                        help='Lancet VCF')
+    parser.add_argument('--vardict_vcf',
+                        help='VarDict VCF')
+    parser.add_argument('--cram', help='CRAM file')
+    parser.add_argument('--output_basename',
+                        help='String to use as basename for output file')
+
+    args = parser.parse_args()
+
+    # Create VCF and CRAM pysam objects
+    strelka2_vcf = pysam.VariantFile(args.strelka2_vcf, 'r')
+    mutect2_vcf = pysam.VariantFile(args.mutect2_vcf, 'r')
+    lancet_vcf = pysam.VariantFile(args.lancet_vcf, 'r')
+    vardict_vcf = pysam.VariantFile(args.vardict_vcf, 'r')
+
+    normal_cram = pysam.AlignmentFile(args.cram, 'rc')
+    
+    # Create output vcf
+    base_dir = os.path.split(os.path.abspath(args.strelka2_vcf))[0]
+    output_vcf_name = build_output_name(args.vardict_vcf, args.output_basename)
+    output_vcf_path = os.path.join(base_dir, output_vcf_name)
+    # print(output_vcf_path)
+    output_vcf = pysam.VariantFile(output_vcf_path, 'w')
+    write_output_header(output_vcf, strelka2_vcf.header.samples, 
+                        strelka2_vcf.header.contigs.values())
+
+    all_variants_dict = {}
+
+    # Read in all variant records
+    for caller, vcf in [('strelka2', strelka2_vcf),
```

Comment:
Maybe could use concurrent.futures to read all four files at the same time? Overall the script ran for about 38 minutes...didn't quite finish, but could be because of a network disconnect. I think we can do better

Reply:
Just to continue bringing the conversation over from Slack, most of the running time is spent getting mapq information from the cram. I think it is possible to parallelize this, but I'm not positive and it would be a small project. My preference would be to do this in a separate PR. Could I ask if this would be ok with you? If so I will create a ticket; but if not I'm happy to work on it here.

Reply:
Yeah, for sure, let's do that!